### PR TITLE
Update EIP-2935: Move to Draft

### PR DESCRIPTION
**Pull request by gballet**
This heretofore stagnant eip is now necessary for stateless clients to be able to execute the `BLOCKHASH` opcode. This PR:

 * moves the eip from a stagnant status to a draft status
 * change how the PR is activated since forks are now timestamp-based
 * adds additional history persist step on fork activation so that new retrieval logic can be transitioned to from first block itself

### CODE REVIEW

File: EIPS/eip-2935.md
Comment (gballet) [inline EIPS/eip-2935.md:4]:
```suggestion
author: Vitalik Buterin (@vbuterin), Tomasz Stanczak (@tkstanczak), Guillaume Ballet (@gballet), Gajinder Singh (@g11tech)
```

File: EIPS/eip-2935.md
Comment (gballet) [inline EIPS/eip-2935.md:5]:
```suggestion
description: store previous block hashes as storage slots of a system contract to allow for stateless execution
author: Vitalik Buterin (@vbuterin), Tomasz Stanczak (@tkstanczak), Guillaume Ballet (@gballet), Gajinder Singh (@g11tech), Tanishq Jasoria (@tanishqjasoria)
```

File: EIPS/eip-2935.md
Comment (gballet) [inline EIPS/eip-2935.md:42]:
```suggestion
Edge cases:

 * For verkle at genesis, no history is written to the genesis state, and at the start of block `1`, genesis hash will be written as a normal operation to slot `0`.
 * for verkle activated at block `1`, only genesis hash will be written at slot `0` as no additional history needs to be persisted.
 * for verkle activated at block `32`, block `31`'s hash will be written to slot `31` and additonal history for `0..30`'s hashes will be persisted, so all in all `0..31`'s hashes.
```

**Review by g11tech — APPROVED** [review]
lgtm pending previous authors' approval

File: EIPS/eip-2935.md
Comment (jochem-brouwer) [inline EIPS/eip-2935.md:37]:
```suggestion
 * **additionally** if this is the first block of the fork, then store last `min(HISTORY_SERVE_WINDOW, block.number)` **blocks' prevhash** into their corresponding slots i.e persist the entire `HISTORY_SERVE_WINDOW` block hashes (up until the genesis) that can be accessed in the block.
```

I think there is an off-by-one error here. If this EIP gets activated at block 0, then at block 1 we get `min(255, 0)` (thus 0) and therefore the "last 0 **blocks' prevhash**" get written to the contract. From other changes to the EIP I get the impression that we want the genesis hash to be written as well.

In general I get the idea that the idea of the BLOCKHASH opcode does not change during the 256 block-window of the fork happening, therefore I think we have this off-by-one-error.

In case of the fork happening at block 256: we would still have access to the `BLOCKHASH` of the blocks 0-255 (so 256 blocks in total), so we need to write the latest 256 blocks (so 0-255) to the contract, not the latest 255 blocks (so 1-255) (`min(256-1, 256-1) = min(255,255) = 255)` per original).

File: EIPS/eip-2935.md
Comment (jochem-brouwer) [inline EIPS/eip-2935.md:43]: Nit: this EIP is not specific to Verkle, so
```suggestion
 * For the fork to be activated at genesis, no history is written to the genesis state, and at the start of block `1`, genesis hash will be written as a normal operation to slot `0`.
```

(also for the other examples)

File: EIPS/eip-2935.md
Comment (jochem-brouwer) [inline EIPS/eip-2935.md:36]: Nit: there is currently no note on how to avoid EIP-158 state clearing. If an account has zero balance, zero nonce, no code, but storage, it is still considered dead, so any touch-action on this contract will delete the entire state. (So either pre-fund this account, or bump the nonce to one, this should be safe since no one should be able to deploy a contract here / have the keys of this contract)

**Review by jochem-brouwer — COMMENTED** [review]
Great updates, I left some comments

File: EIPS/eip-2935.md
Comment (g11tech) [reply EIPS/eip-2935.md:37]: this step is previous to additional step, so block 1 writes its prevhash i.e. 0 => genesis hash, and then no additional things get written

File: EIPS/eip-2935.md
Comment (g11tech) [inline EIPS/eip-2935.md:44]:
```suggestion
 * for activation at block `1`, only genesis hash will be written at slot `0` as no additional history needs to be persisted.
```

File: EIPS/eip-2935.md
Comment (g11tech) [inline EIPS/eip-2935.md:45]:
```suggestion
 * for activation at block `32`, block `31`'s hash will be written to slot `31` and additonal history for `0..30`'s hashes will be persisted, so all in all `0..31`'s hashes.
```

File: EIPS/eip-2935.md
Comment (jochem-brouwer) [reply EIPS/eip-2935.md:37]: Yup, I missed the `additional` point

Comment (gballet) [reply EIPS/eip-2935.md:37]: it's correct, because in the general case, you have to add 255 blocks after adding the one just before (notice the **additionally** in bold).

In the case of genesis, you also get `min(255, 0) == 0` so no extra block get inserted.

I think we should switch to a pseudo-python implementation, because you're the second person to be confused by this (I was the first :D)

Comment (gballet) [reply EIPS/eip-2935.md:37]: so something like:

```python3
def process_block_hash_history(block :Block, state: State):
    if block.timestamp >= FORK_TIMESTAMP:
        state.insert_slot(HISTORY_STORAGE_ADDRESS, block.number-1, block.parent.hash)

    # If this is the first fork block, add the parent's direct 255 ancestors as well
    if block.parent.timestamp < FORK_TIMESTAMP:
        ancestor = block.parent
        for i in range(255):
            ancestor = ancestor.parent
            state.insert_slot(HISTORY_STORAGE_ADDRESS, ancestor.number, ancestor.hash)
```

File: EIPS/eip-2935.md
Comment (jochem-brouwer) [inline EIPS/eip-2935.md:34]:
```suggestion
At the start of processing any block where `block.timestamp >= FORK_TIMESTAMP` (ie. before processing any transactions), update the history in the following way:
```

Current text does not match pseudocode

File: EIPS/eip-2935.md
Comment (jochem-brouwer) [inline EIPS/eip-2935.md:53]:
```suggestion
Note that if this is the fork block, then it persists the additional requisite history that could be needed while resolving `BLOCKHASH` opcode for all of the `HISTORY_SERVE_WINDOW` ancestors (up until genesis).
```

File: EIPS/eip-2935.md
Comment (jochem-brouwer) [inline EIPS/eip-2935.md:55]:
```suggestion
For resolving the `BLOCKHASH` opcode this fork onwards (`block.timestamp >= FORK_TIMESTAMP`), switch the logic to:
```

File: EIPS/eip-2935.md
Comment (jochem-brouwer) [inline EIPS/eip-2935.md:59]: This does not match current `BLOCKHASH`. If `arg >= block.number` then return 0 (push 0 to stack)

Our code:

```typescript
      const number = runState.stack.pop()

      const diff = runState.interpreter.getBlockNumber() - number
      // block lookups must be within the past 256 blocks
      if (diff > BIGINT_256 || diff <= BIGINT_0) {
        runState.stack.push(BIGINT_0)
        return
      }
```

**Review by jochem-brouwer — CHANGES_REQUESTED** [review]

File: EIPS/eip-2935.md
Comment (gballet) [reply EIPS/eip-2935.md:59]: yeah but that's precisely why we want to use the contract: be able to get further history than 256 blocks.

Comment (jochem-brouwer) [reply EIPS/eip-2935.md:59]: Yes, the point is about when you request the current block number hash, or a block number higher, I would assume that `assert` throws? (And thus push nothing to stack and exit call frame?)

File: EIPS/eip-2935.md
Comment (jochem-brouwer) [inline EIPS/eip-2935.md:76]:
```suggestion
While the second option looks more elegant, it has a higher complexity as well as gas consumption considerations.
```

File: EIPS/eip-2935.md
Comment (g11tech) [reply EIPS/eip-2935.md]: applied the feedback

File: EIPS/eip-2935.md
Comment (jochem-brouwer) [inline EIPS/eip-2935.md:61]: In EVM, the requested `block.number` cannot be negative or be higher than 2**256, so just the check `arg >= block.number` (assuming that `block` is the current block we are running on) is sufficient here.

Comment (jochem-brouwer) [reply EIPS/eip-2935.md:61]: `arg` should thus also be of type `uint256`

Comment (g11tech) [reply EIPS/eip-2935.md:61]: yes we need to discuss the wrap around thing and arg sizes

File: EIPS/eip-2935.md
Comment (s1na) [inline EIPS/eip-2935.md:30]: This conflicts with the system address from EIP-4788.

File: EIPS/eip-2935.md
Comment (gballet) [inline EIPS/eip-2935.md:49]: @g11tech @jochem-brouwer this seems incorrect, as if the fork happens at block 1, the genesis will not be stored, either by this method, or by the regular block operation. I think this check should be moved AFTER `state.insert_slot`

Comment (jochem-brouwer) [reply EIPS/eip-2935.md:49]: Why? If fork block is 1, then 

```
    if block.timestamp >= FORK_TIMESTAMP:
        state.insert_slot(HISTORY_STORAGE_ADDRESS, (block.number-1) % HISTORY_SERVE_WINDOW , block.parent.hash)
```

This will be ran (so genesis hash is stored).

File: EIPS/eip-2935.md
Comment (gballet) [inline EIPS/eip-2935.md:47]: The wording here is confusing, I suggest the following:

```suggestion
            #  if ancestor.number == 0, this means we inserted the
            # genesis block hash in the previous iteration. Stop here.
```

File: EIPS/eip-2935.md
Comment (g11tech) [reply EIPS/eip-2935.md:49]: i think L51 makes one hop to the parent,so at number=1, L51 will hop to genesis block and save it, may be we can simplify it

File: EIPS/eip-2935.md
Comment (g11tech) [reply EIPS/eip-2935.md:52]: this makes it look more analogous to L41

Comment (gballet) [reply EIPS/eip-2935.md:52]: actually, I think we should just rewrite it more clearly:

```python
for i in range(block.number-1, min(block.number-HISTORY_SERVE_WINDOW, 0)):
    state.insert_slot(HISTORY_STORAGE_ADDRESS, i % HISTORY_SERVE_WINDOW, ancestor.parent.hash)
    ancestor = ancestor.parent
```

then, the way it's implemented is left to the implementer.

Comment (g11tech) [reply EIPS/eip-2935.md:52]: i think this is correct  storing at (`i-1`) than `i`, since 'i` is ancestor

or we shouldn't use i, just `ancestor.number -1` since that relationship is more clear (we always store ancestor's parent)

```python
    state.insert_slot(HISTORY_STORAGE_ADDRESS, (ancestor.number -1) % HISTORY_SERVE_WINDOW, ancestor.parent.hash)
```

**Comment by lightclient** [comment]
@g11tech @gballet how are the block hashes written to this contract? The assembly I see has no `sstore`. I think it should follow the 4788 pattern where it is spec'd as the system makes a call as `0xff..ff` and the update logic is gated on that.

**Review by tkstanczak — APPROVED** [review]

**Review by lightclient — APPROVED** [review]

**Comment by gballet** [comment]
> @g11tech @gballet how are the block hashes written to this contract? The assembly I see has no `sstore`. I think it should follow the 4788 pattern where it is spec'd as the system makes a call as `0xff..ff` and the update logic is gated on that.

We already addressed this question during ACD: if anything, the 4788 system should be changed as it unnecessarily adds code chunks to the witness. We do not want to do that, and so the 2935 contract retains its original mode of operation

**Pull request merged**
